### ISSUE TITLE
How to understand the sphericalMapping parameters?

### ISSUE THREAD

**Issue by rockywind** [astra-vision/SceneRF#4]
Hi,
How to set the magical number, such as 104.7294 in our dataset?
![image](https://user-images.githubusercontent.com/14838781/210505504-486719e2-c7c9-4127-ac38-5a903e169034.png)

**Comment by anhquancao** [comment]
Hi,
You can take a look how the v_angle and h_angle are computed here:
https://github.com/astra-vision/SceneRF/blob/main/scenerf/models/spherical_mapping.py#L98
To get their max value, you simply get the v_angle and h_angle of all the pixels in the image then compute the min and max

**Comment by rockywind** [comment]
thanks a lot.
How to set the add_fov_ver=8 and add_fov_hor=20?
![image](https://user-images.githubusercontent.com/14838781/210530650-1ec9a95a-1c9f-490a-9fe4-7bb861fdc2bc.png)

**Comment by rockywind** [comment]
I used the fisheye image to train the model, but the render rgb image is fuzzy. Just like the image below.
![image](https://user-images.githubusercontent.com/14838781/210531063-f471dc7c-31d7-44f8-831b-dd0c8289a516.png)
This is the gt rgb image.
![image](https://user-images.githubusercontent.com/14838781/210531199-425eac9c-b7e1-4aff-8c76-fb44db7e9e45.png)

**Comment by anhquancao** [comment]
The vertical FOV is: 104 - 75 = 29 deg
The horizontal FOV is: 131 - 49 = 82 deg
Since we want to enlarge by 150%, we add 16 deg and 40 deg to the vertical and horizontal angles, resulting in 16/2 and 40/2 angles added to each side.

Note that I consider the rays in the image coordinate of a perspective camera. You might need to consider different formula to compute the rays for fisheye camera. Also, the point coordinates are in the pixel coordinates also.

**Comment by rockywind** [comment]
I added the fisheye distortion from camera coordinate to image coordinate.
![image](https://user-images.githubusercontent.com/14838781/210531651-4b2ba37e-2b35-47c6-838e-5ad3460e6cf1.png)

**Comment by anhquancao** [comment]
Can you try to overfit on a single image to see? The image seems to be distorted in the other way

**Comment by anhquancao** [comment]
> I added the fisheye distortion from camera coordinate to image coordinate. ![image](https://user-images.githubusercontent.com/14838781/210531651-4b2ba37e-2b35-47c6-838e-5ad3460e6cf1.png)

Where you put this line? How did you modify the ray sampling? like in here
https://github.com/astra-vision/SceneRF/blob/main/scenerf/models/utils.py#L56

**Comment by anhquancao** [comment]
Can you try the mapping on a single image to see?
Before spherical projection:
![before](https://user-images.githubusercontent.com/9668858/210534191-6044c636-b1e5-4b55-a536-cdaf36810e53.jpg)
After spherical projection:
![after](https://user-images.githubusercontent.com/9668858/210534174-30cbc436-da73-4f80-a57c-821339e1b1d1.jpg)

**Comment by anhquancao** [comment]
You can try smth like this code:
```
if __name__ == "__main__":
    # cam_K = torch.tensor([[707.0912,   0.0000, 601.8873],
    #                       [0.0000, 707.0912, 183.1104],
    #                       [0.0000,   0.0000,   1.0000]])
    cam_K = torch.tensor([[518.8579, 0, 320],
                          [0, 518.8579, 240],
                          [0.0000,   0.0000,   1.0000]])
    inv_K = torch.inverse(cam_K)
    T_velo2cam = torch.tensor([[[-0.0019, -1.0000, -0.0080,  0.0615],
                                [-0.0065,  0.0081, -0.9999, -0.0734],
                                [1.0000, -0.0018, -0.0065, -0.3340],
                                [0.0000,  0.0000,  0.0000,  1.0000]]])

    T_cam2velo = torch.inverse(T_velo2cam)

    # img_W, img_H = 1220, 370
    img_W, img_H = 640, 480

    out_img_W = 1650 
    out_img_H = 500
    mapping = SphericalMapping(
        v_angle_max=104.7294 + 8,
        v_angle_min=75.4815 - 8,
        h_angle_max=131.1128 + 20,
        h_angle_min=49.5950 - 20,
        # v_angle_max=104.7294,
        # v_angle_min=75.4815,
        # h_angle_max=131.1128,
        # h_angle_min=49.5950,
        # fov_up=9 + 12,
        # fov_down=-16 - 12,
        # fov_left=-33 - 34,
        # fov_right=36 + 34,
        img_W=img_W, img_H=img_H, out_img_W=out_img_W, out_img_H=out_img_H)
    # proj_x_1, proj_y_1 = mapping.from_pixels(inv_K=ing_K, depth_anchor=50)

    pix_coords = torch.tensor([
        [100, 100],
        # [100, 200],
        # [100, 300],
        [-1, 100],
        [1221, 100],
        [100, 371],
        [100, -1],
    ]).float()
    
    # pix_coords = None
    pix_coords, pix_sphere_coords, distance = mapping.from_pixels(
        inv_K=inv_K, pix_coords=None)
    pdb.set_trace()
    
    kitti_root = "/gpfswork/rech/kvd/uyl37fq/data/semantic_kitti"
    kitti_preprocess_root = "/gpfsscratch/rech/kvd/uyl37fq/monoscene_preprocess/kitti"
    write_path = "/gpfsscratch/rech/kvd/uyl37fq/temp/draw_output/monoscene2/dataloader"

    ds = KittiDataset_18_8(
        split="val",
        root=kitti_root,
        img_scale=1,
        preprocess_root=kitti_preprocess_root,
        sequence_distance=10,
        frames_interval=0.4,
        max_sample_depth=100,
        fliplr=0,
        color_jitter=None,
        reliable_depth=52,
        n_rays=1024
    )
    x = ds[2]['img_sources'][0]

    img = x.permute(1, 2, 0).numpy()
    plt.imsave("before.jpg", img)
    x = x.unsqueeze(0)
    # scale = 1
    # out_W, out_H = out_img_W//scale, out_img_H//scale
    # map_sphere = torch.zeros((out_W, out_H, 2)).type_as(x) - 1
    # pix_sphere_scale = (pix_sphere_coords / scale).long()

    # pix_scale = (pix_coords / scale)
    # map_sphere[pix_sphere_scale[:, 0], pix_sphere_scale[:, 1], :] = pix_scale
    # map_sphere = map_sphere.reshape(-1, 2)

    # map_sphere[:, 0] /= x.shape[3]
    # map_sphere[:, 1] /= x.shape[2]

    # map_sphere = map_sphere * 2 - 1
    # map_sphere = map_sphere.reshape(1, 1, -1, 2)

    # feats = F.grid_sample(
    #     x,
    #     map_sphere,
    #     align_corners=False,
    #     mode='bilinear',
    #     padding_mode="zeros"
    # )
    # feats = feats.reshape(feats.shape[0], feats.shape[1], out_W, out_H)
    # feats = feats.permute(0, 1, 3, 2)
    scale = 2
    x = F.interpolate(
            x,
            size=(x.shape[2]//scale, x.shape[3]//scale),
            mode="bilinear",
            align_corners=True,
        )
    pdb.set_trace()
    feats = get_sphere_feature(x, pix_coords, pix_sphere_coords, scale, out_img_W, out_img_H)
    pdb.set_trace()
    img = feats.squeeze(0).permute(1, 2, 0).numpy()
    plt.imsave("after.jpg", img)
    # pdb.set_trace()
```

**Comment by rockywind** [comment]
Hi，
Thanks for your help!
I can't find the KittiDataset_18_8 function. Can you show me the detailed code?
![image](https://user-images.githubusercontent.com/14838781/210755019-06513b9a-d7ad-4c1d-8f53-8ab269112729.png)

**Comment by anhquancao** [comment]
It's just the KittiDataset but unclean, it only used to get an image: x = ds[2]['img_sources'][0]

**Comment by rockywind** [comment]
Hi,
I used the spherical projection code in kitti dataset, but there are large black region  after spherical projection.
Before spherical projection
![image](https://user-images.githubusercontent.com/14838781/210804374-d3e41e8e-2f81-47a1-a420-7dce4050a1ad.png)

After spherical projection
![image](https://user-images.githubusercontent.com/14838781/210804440-bb06da64-59e8-4cb0-a021-d78e44e1a8f8.png)

 This is my code
````
import matplotlib.pyplot as plt
import torch.nn.functional as F
import torch
from scenerf.models.spherical_mapping import get_sphere_feature
from scenerf.data.semantic_kitti.kitti_dataset import KittiDataset as KittiDataset_18_8
from scenerf.models.spherical_mapping import SphericalMapping
if __name__ == "__main__":
    cam_K = torch.tensor([[707.0912,   0.0000, 601.8873],
                          [0.0000, 707.0912, 183.1104],
                          [0.0000,   0.0000,   1.0000]])
    # cam_K = torch.tensor([[518.8579, 0, 320],
    #                       [0, 518.8579, 240],
    #                       [0.0000,   0.0000,   1.0000]])
    inv_K = torch.inverse(cam_K)
    # T_velo2cam = torch.tensor([[[-0.0019, -1.0000, -0.0080,  0.0615],
    #                             [-0.0065,  0.0081, -0.9999, -0.0734],
    #                             [1.0000, -0.0018, -0.0065, -0.3340],
    #                             [0.0000,  0.0000,  0.0000,  1.0000]]])
    T_velo2cam = torch.tensor([[ 4.27680239e-04, -9.99967248e-01, -8.08449168e-03,5.11466120e-02],
                                [-7.21062651e-03,  8.08119847e-03, -9.99941316e-01, -5.40398473e-02],
                                [ 9.99973865e-01,  4.85948581e-04, -7.20693369e-03,-2.92196865e-01],
                                [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,1.00000000e+00]])
    T_cam2velo = torch.inverse(T_velo2cam)

    img_W, img_H = 1220, 370

    out_img_W = 1500
    out_img_H = 452
    add_fov_ver = 8
    add_fov_hor = 20
    mapping = SphericalMapping(
            v_angle_max=104.7294 + add_fov_ver,
            v_angle_min=75.4815 - add_fov_ver,
            h_angle_max=131.1128 + add_fov_hor,
            h_angle_min=49.5950 - add_fov_hor,
            img_W=img_W, img_H=img_H, out_img_W=out_img_W, out_img_H=out_img_H)


    pix_coords, pix_sphere_coords, distance = mapping.from_pixels(
        inv_K=inv_K, pix_coords=None)
    # pdb.set_trace()
    
    # kitti_root = "/SHFP12/02_bevdet/SceneRF/own_data/data_odometry/"
    # kitti_preprocess_root = "/SHFP12/02_bevdet/SceneRF/own/preprocess"
    # write_path = "/SHFP12/02_bevdet/SceneRF/own_data_log_02/draw_out"
    kitti_root = "/SHFP12/02_bevdet/SceneRF/kitti_odometry"
    kitti_preprocess_root = "/SHFP12/02_bevdet/SceneRF/kitti/preprocess"
    write_path = "/SHFP12/02_bevdet/SceneRF/log_kitti/draw_out"

    ds = KittiDataset_18_8(
        split="val",
        root=kitti_root,
        preprocess_root=kitti_preprocess_root,
        sequence_distance=10,
        n_sources=1,
        eval_depth=80,
        n_rays=1024
    )
    x = ds[2]['img_sources'][0]

    img = x.permute(1, 2, 0).numpy()
    plt.imsave("before.jpg", img)
    x = x.unsqueeze(0)

    scale = 1
    x = F.interpolate(
            x,
            size=(x.shape[2]//scale, x.shape[3]//scale),
            mode="bilinear",
            align_corners=True,
        )

    feats = get_sphere_feature(x, pix_coords, pix_sphere_coords, scale, out_img_W, out_img_H)
    img = feats.squeeze(0).permute(1, 2, 0).numpy()
    plt.imsave("after_2x.jpg", img)

````

**Comment by anhquancao** [comment]
It's correct, the black region is because of the increase in the FOV by ~150%. You can try different  add_fov_ver and add_fov_hor values

**Comment by rockywind** [comment]
Thanks.
I try it!

**Comment by rockywind** [comment]
Hi,

Why there are black curve line after spherical orijection?
````
    out_img_W = 1500
    out_img_H = 452
    add_fov_ver = 1
    add_fov_hor = 1 #20
    mapping = SphericalMapping(
            v_angle_max=104.7294 + add_fov_ver,
            v_angle_min=75.4815 - add_fov_ver,
            h_angle_max=131.1128 + add_fov_hor,
            h_angle_min=49.5950 - add_fov_hor,
            img_W=img_W, img_H=img_H, out_img_W=out_img_W, out_img_H=out_img_H)
````
After spherical projection
![image](https://user-images.githubusercontent.com/14838781/210805994-15311562-c10a-4a43-93ef-b7a72fbbb3c3.png)

**Comment by anhquancao** [comment]
Because you map the original image to too big image plane. The mapping is just to map a pixel from the original image to a pixel in the mapped image. 
Btw, I recommend you try w/o the spherical mapping by replacing it with the Planar Mapping
```
class PlanarMapping(nn.Module):

    def __init__(self, 
        img_W, img_H, 
        out_img_W, out_img_H, 
        fov_increase=1.5):
        super(PlanarMapping, self).__init__()
        self.img_W = img_W
        self.img_H = img_H
        self.out_img_W = out_img_W
        self.out_img_H = out_img_H
        self.fov_increase = fov_increase
        print("==> mapping version Oct 10")

        
    def from_pixels(self, device=None, pix_coords=None):
        if pix_coords is None:
            meshgrid = np.meshgrid(
                range(self.img_W), range(self.img_H), indexing='xy')
            id_coords = np.stack(meshgrid, axis=0).astype(np.float32)
            id_coords = torch.from_numpy(id_coords)
            pix_coords = torch.cat(
                [id_coords[0].reshape(-1, 1), id_coords[1].reshape(-1, 1)], 1).to(device)
        
        pix_out_coords = torch.zeros_like(pix_coords)
        pix_out_coords[:, 0] = ((pix_coords[:, 0] / self.img_W) / self.fov_increase) * self.out_img_W + (self.out_img_W/self.fov_increase) * 0.25
        pix_out_coords[:, 1] = ((pix_coords[:, 1] / self.img_H) / self.fov_increase) * self.out_img_H + (self.out_img_H/self.fov_increase) * 0.25
        # pdb.set_trace()
        return pix_coords, pix_out_coords
```

**Comment by rockywind** [comment]
Thank you very much!
I try it again!

**Comment by rockywind** [comment]
Hi,
There are the plane map and the spherical map in our dataset.
I have 3 questions below.
Can you help me?
1. I was confused that the plane map is very similar to the original image.
2.  I never add the fisheye distortion in spherical projection, but it seems no fluence.
3. Why do we use the spherical projection in SceneRF rather than the plane map?
Before
![before](https://user-images.githubusercontent.com/14838781/210827321-39c77eba-f4d7-4c58-878a-8e6cc104fc66.jpg)
After plane projection
![after_own_planemap](https://user-images.githubusercontent.com/14838781/210827409-f5f79d1b-e290-4268-ac3f-e6b01000f54d.jpg)
After spherical projection
![after_own_spherical](https://user-images.githubusercontent.com/14838781/210827508-67d7c918-0bf7-43e7-a105-df0651b4fb86.jpg)

**Comment by anhquancao** [comment]
Please check the Sec. 3.3 in the paper. 
The idea is to create a larger image plane which we can run convolution to provide features to sampled points outside the image FOV.  Because the rays pass through the pixels in the perspective image, often go outside the FOV (the scenery observed by the image).
The assumption is that the input image is a perspective image.

**Comment by anhquancao** [comment]
I recommend you to try without any mapping first and just try to overfit a single image to see

**Comment by rockywind** [comment]
Thank you very much!
I try it!

**Comment by rockywind** [comment]
Hi,
I trained single image without any mapping in 100epoch.
Render color:000001_0.0_0.png
![image](https://user-images.githubusercontent.com/14838781/211009005-00d77f80-384e-4a1e-a678-2129550c2615.png)
Depth_visual:000001_0.0_0.png
![image](https://user-images.githubusercontent.com/14838781/211009101-39ef40fd-14b3-4f3c-ada2-04d6405891bb.png)
Loss curve
![image](https://user-images.githubusercontent.com/14838781/211009459-79f3385c-4ad4-4444-a55d-945637915d8f.png)
![image](https://user-images.githubusercontent.com/14838781/211009510-cb866cc0-236a-48e5-8fdd-78bca60a94bc.png)
This is my code without any mapping.
![image](https://user-images.githubusercontent.com/14838781/211009761-7994dbad-1deb-4314-9a62-fdd29bc9e992.png)

**Comment by anhquancao** [comment]
Sorry, I have no idea! You need to check:
1.  The points sampling 
2.  Ray rendering process, especially the points sampled on rays and
3.  How points are projected on image since it is designed for perspective image. 
4. Furthermore, please check your pose transformation as well to see if the sampled points on source frame coordinates are transformed correctly to the infer frame coordinates.

**Comment by rockywind** [comment]
Thank a lot!

**Comment by anhquancao** [comment]
Hi @rockywind ,
I have included a script for calculating these figures at the following link:
https://github.com/astra-vision/SceneRF/blob/main/scenerf/scripts/determine_angles.py
To execute this code, you may use the following command: python scenerf/scripts/determine_angles.py

**Comment by rockywind** [comment]
Hi @anhquancao ,
Thank you very much!
